File: dotnet/farm/Farm.cs
using System;
using System.Collections.Generic;
namespace dotnet
{
    ///<sumary>
    /// Author: brayvasq <brayvasq@gmail.com>
    /// Version: 0.1
    /// Date: 2018-03-25
    /// This class just simulate a Farm.
    ///</sumary>
    public class Farm
    {
        public string name {get; set;}
        public string address {get; set;}
        public List<Animal> animals;
        ///<sumary>
        /// This methods is the constructor of the class.
        /// The attribute animals contains all animals of the farm.
        ///</sumary>
        ///<param name="name">is a identifier of the farm.</param>  
        ///<param name="address">is a simple phisical direction of the farm.</param>         
        public Farm(string name,string address){
            this.name = name;
            this.address = address;
            this.animals = new List<Animal>();
        }
        ///<sumary>This Method just show the information of the farm</sumary>
        public void getInfo(){
            Console.WriteLine("Name : "+this.name+"\nAddress : "+this.address);
        }
        ///<sumary>This method iterate the list of animals and show basic info of each animal</sumary>
        public void getAnimalsInfo(){
            if(this.animals.Count>0){
                foreach (var i in this.animals)
                {
                    Console.WriteLine("Id : "+i.id+" - Name : "+i.name);
                }
            }
        }
        ///<sumary>This method iterate the list of animals for search a specific animal</sumary>
        ///<param name="id">is the identifier of the animal to search</param>          
        ///<returns>resp, contain the animal with the searched id</returns>
        public Animal searchAnimal(string id){
            Animal resp = null;
            foreach(var i in this.animals){
                if(i.id.Equals(id)){
                    resp = i;
                }
            }
            return resp;
        }
        ///<sumary>This method just add a new animal to the list of animals</sumary>
        ///<param name="animal">is a object that will be added to the list</param>                  
        public void addAnimal(Animal animal){
            this.animals.Add(animal);
        }
        ///<sumary>This method delete a specific animal based on a id</sumary>
        ///<param name="id">the id of animal to search and delete</param>                          
        public void delAnimal(string id){
            Animal animal = searchAnimal(id);
            try{
                this.animals.Remove(animal);
            }catch(Exception e){Console.WriteLine("Error"+e);}
        }
        ///<sumary>
        /// This method implements the algorithm bubble sort
        /// To sort the list of animals based on his price
        ///</sumary>
        ///<returns>response, is a list with the animals sorted by price<returns>
        public List<Animal> sortByPrice(){
            List<Animal> response = this.animals;
            int limit = response.Count;
            if(limit>1){
                for(int i=1;i<limit;i++){
                    for(int j=0;j<(limit-i);i++){
                        if(response[j].price > response[j+1].price){
                            Animal aux = response[j];
                            response[j] = response[j+1];
                            response[j+1] = aux;
                        }
                    }
                }
            }
            return response;
        }
    }
}

File: kotlin/src/main/farm/Farm.kt
package farm
import animals.Animal

/**
 * @author brayvasq <brayvasq@gmail.com>
 * @version 0.1
 * @since 2018-03-25
 * This class just simulate a Farm.
 * @param name is a identifier of the farm.
 * @param address is a simple phisical direction of the farm.
 * the attribute animals contains all animals of the farm.
 */
class Farm(var name:String,var address:String){

    var animals : ArrayList<Animal> = ArrayList()

    /**
     * This Method just show the information of the farm
     */
    fun getInfo() = println("Name : ${this.name}\nAddress : ${this.address}")

    /**
     * This method get the number of animals in the list of animals.
     * @return lenght of the list of animals.
     */
    fun getNumberAnimals() : Int = animals.size

    /**
     * This method iterate the list of animals and show basic info of each animal
     */
    fun getAnimalsInfo(){
        if (!this.animals.isEmpty()) {
            for (i in this.animals) {
                println("Id : " + i.id + " - Name : " + i.name)
            }
        } else {
            print("List of animals is empty")
        }
    }

    /**
     * This method iterate the list of animals for search a specific animal
     * @param id is the identifier of the animal to search
     * @return resp, contain the animal with the searched id
     */
    fun searchAnimal(id: String): Animal? {
        var resp: Animal? = null
        
        for (e in this.animals) {
            if (e.id == id) {
                resp = e
            }
        }

        return resp
    }

    /**
     * This method get the sum of the price of the all animals in the list
     * @return price, is the total price.
     * */
    fun getTotalPrice() : Double{
        var price:Double = 0.0
        if(!this.animals.isEmpty()){

            val iterator = this.animals.iterator()
            
            iterator.forEach {
                price += it.price
            }

        }

        return price
    }

    /**
     * This method just add a new animal to the list of animals
     * @param animal is a object that will be added to the list
     */
    fun addAnimal(animal: Animal) {
        this.animals.add(animal)
    }

    /**
     * This method delete a specific animal based on a id
     * @param id the id of animal to search and delete
     */
    fun delAnimal(id: String) {
        val animal = searchAnimal(id)

        try {
            this.animals.remove(animal)
        } catch (e: Error) {
            println("Error")
        }
    }

    /**
     * This method implements the algorithm bubble sort
     * to sort the list of animals based on his price
     * @return response, is a list with the animals sorted by price
     */
    fun sortByPrice(): ArrayList<Animal> {
        val response = this.animals
        val limit = response.size

        if (limit > 1) {
            for (i in 1 until limit) {
                for (j in 0 until limit - i) {
                    if (response[j].price > response[j + 1].price) {
                        val aux = response[j]
                        response[j] = response[j + 1]
                        response[j + 1] = aux
                    }
                }
            }
        }

        return response
    }

    /**
     * This method sort the list of animals based on his age.
     * @return list of animals sort by age.
     * */
    fun sortByAge() : ArrayList<Animal>{
        var sortedList : ArrayList<Animal> = (ArrayList(this.animals.sortedWith(compareBy({ it.age }))))
        return sortedList
    }
}